LEAN-3301

mnatsheh opened the pull request.

mbartenev-atypon commented:
In general this approach is not something what I see to be advantageous and this is for several reason:

1. We should never encode models/nodes individually. There is no assurance that encoding and decoding process will not affect any other nodes or these nodes will not be affected. It may be alright for these nodes but there is no architectural enforcement that would guarantee that the assumption on which this logic is based will always be correct. This is why in adaptTrackedData in the article editor we encode/decode the entire document and cherry pick needed parts. This would also makes further development more demanding as we need to remember that we encode current document not only in the article-editor but also for some nodes in the body editor.

2. Also related to adaptTrackedData - nodes do not only need encoding/decoding but trackedData need separate handling, that we provide in that function (excluding deleted nodes, finding actual current attributes values). Please see src/tracked-models/adapt-tracked-data.ts in the article-editor for more details. Maybe I'm missing something but I don't see it here.

3. We mix up layers of architecture extensively. Views get access to the encoding/decoding. Whereas the body editor never handled data transforms before and this should stay that way. There was an exception related to bibliographies but ideally we should've cleaned that up instead of adapting it as a routine approach. In short, ideally body-editor shouldn't decode or encode data.

4. This makes body-editor aware of the json-schema, and also may vet further use of it, which is likely because in certain cases it's easier to work with json-schema instead of PMDoc.

5. There is a lot of logic on the side of the body editor to cater for the author modals. As I see it, this shouldn't be the body-editor's responsibility to serve other visual components for their needs in such extent, which happens here because authors modals rely on a different data model.

It feels like this approach is a bit undermined by not converting author modals to actually rely on prosemirror document, which, as I commented on LEAN-3083 PR, is something what needs to be done to transfer the modals here. What happens here is that we move authors modals out of the article-editor but they are still completely json-schema dependent, but now instead of going through the main process of preparing data in the article-editor we process nodes here separately, which makes things more spread out and less clear and puts a contributors node view in charge of a very responsible process of data conversion.

I see 2 possible solutions here:
1. Reimplement author nodes so they rely on the PM nodes as on data instead of JSON-schema. As I mentioned on LEAN-3083 PR that seems to be a significant effort and it would also need to be developed with track-changes in mind.
2. Refactor and cleanup in such way that author-modals will rely on JSON-schema produced by adaptTrackedData (trackedModelMap) in the article editor. While that may mean accommodation of the modals in the article-editor, this offers an advantage of abstracting track-changes related logic from all the react components, such as inspector and author-modals as well.

Both seem ok to me: Article-editor role as a broker between json-schema and PMDoc will not go away in any case so we can take advantage of that. Second approach would be perhaps better if we are ever to get rid of json-schema.

asouqi commented:
@mbartenev-atypon can we move `adaptTrackedData` to style-guide so we could use it directly here for: (figures dropdown list, bibliography ..)? as I tried to pass `trackModelMap` to body-editor but that didn't work well as in some cases the data kept steel

mnatsheh commented on package.json:
```diff
@@ -1,7 +1,7 @@
 {
```
To me, the (confusingly named) json-schema module represents the types that our system supports, so if someone wants to work on contributors, they can use the `Contributor` type, and similarly for references and `BibliographyItem`. I don't see a problem of being json-schema dependent; the whole point of the module is that.

Whether or not it's appropriate to use encode/decode here, I would like for the time being to keep ProseMirror-related logic contained within body-editor, and any external communication should happen through json-schema, so a component in the style-guide should not work with PM.

> We should never encode models/nodes individually. There is no assurance that encoding and decoding process will not affect any other nodes or these nodes will not be affected.

Why is there no guarantee? Do you know of any case where we actually update the nodes? If so, we should remove that logic, I can't think of any case where we would want the input node to be updated.

> It feels like this approach is a bit undermined by not converting author modals to actually rely on prosemirror document, which, as I commented on LEAN-3083 PR, is something what needs to be done to transfer the modals here. What happens here is that we move authors modals out of the article-editor but they are still completely json-schema dependent, but now instead of going through the main process of preparing data in the article-editor we process nodes here separately, which makes things more spread out and less clear and puts a contributors node view in charge of a very responsible process of data conversion.

I disagree with that, using the store for data communication is way more confusing and more spread out. Currently if I wanted to see what data is being used in the modal, I would have to:
- Go the node view to find out the component is actually passed in props.
- Go to the article-editor to find what the component being passed is.
- Go to the component to find out where the data is coming from.
- Go to the store init logic to find out how this data is managed.

mnatsheh replied:
I also think adapt-tracked-data should be in body-editor, where it can also be used in there.

mbartenev-atypon replied:
> To me, the (confusingly named) json-schema module represents the types that our system supports, so if someone wants to work on contributors, they can use the `Contributor` type, and similarly for references and `BibliographyItem`. I don't see a problem of being json-schema dependent; the whole point of the module is that.

To clarify, when I wrote about JSON-schema I mean data set and not the types - that is the model map that we get by decoding the document. The problem with this is that for almost every json-schema based type we have a peer type in the prosemirror document to use both no the same level is going to be confusing, would it not?


> Why is there no guarantee? Do you know of any case where we actually update the nodes? If so, we should remove that logic, I can't think of any case where we would want the input node to be updated.

It was never designed to work that way: some nodes may reach to other nodes for additional info. For example ref ids (all rid), section priorities are dependent on the general document context. It would be perfectly acceptable if a node would need to reach out to its peer or parent during encoding or decoding, we also have some nodes for which there are no peer json-schema models at all. 



> I disagree with that, using the store for data communication is way more confusing and more spread out. Currently if I wanted to see what data is being used in the modal, I would have to:
> 
> * Go the node view to find out the component is actually passed in props.
> * Go to the article-editor to find what the component being passed is.
> * Go to the component to find out where the data is coming from.
> * Go to the store init logic to find out how this data is managed.

Perhaps it may be somewhat confusing but it is definitely not spread out, since data handling happens in a single place.
- Not sure why we need to pass any components. We don't pass any components as things are now. 
- Not clear why you need to go the component to find out where the data are coming from if they always come only from trackedModelMap. It is rather in this change where things will become less clear as it creates an adhoc case and you would need go searching where exactly data are created in this specific case.
- If by store init logic you mean the complicated approach of building author and affiliations data, then I agree it's something what could've been improved but it doesn't need the modal to be rendered and managed through a node view to be improved.


> Whether or not it's appropriate to use encode/decode here, I would like for the time being to keep ProseMirror-related logic contained within body-editor, and any external communication should happen through json-schema, so a component in the style-guide should not work with PM.

While I agree with that we need to contain prosemirror related logic to the body-editor, even if we move adaptTrackData here, still, we will have to import it to the article-editor, because operations on trackedModelMap will require it and as I see it, it should be used only there. Also article-editor will still have to deal with the doc and it has a lot of prosemirror related logic besides adaptTrackedData: in inspector, in snapshots handling, and in track changes. It also receives the PM doc from api and initiates the body-editor, basically acting as a controller for the body-editor. So while it is possible to abstract json-schema away from the body-editor it doesn't seem likely for the article-editor to be unaware of prosemirror doc and its parts.

In my opinion, If we prefer to revise how we manage data relations between the article-editor and body-editor, we should improve on that and take advantage of that improvement on the application level rather than dealing with this on case by case basis.

mnatsheh replied:
> It was never designed to work that way: some nodes may reach to other nodes for additional info. For example ref ids (all rid), section priorities are dependent on the general document context. It would be perfectly acceptable if a node would need to reach out to its peer or parent during encoding or decoding, we also have some nodes for which there are no peer json-schema models at all.

That's fair. You're saying that `encode` might need data from somewhere else in the document, not that `encode` might update the nodes.

> While I agree with that we need to contain prosemirror related logic to the body-editor, even if we move adaptTrackData here, still, we will have to import it to the article-editor, because operations on trackedModelMap will require it and as I see it, it should be used only there. Also article-editor will still have to deal with the doc and it has a lot of prosemirror related logic besides adaptTrackedData: in inspector, in snapshots handling, and in track changes. It also receives the PM doc from api and initiates the body-editor, basically acting as a controller for the body-editor. So while it is possible to abstract json-schema away from the body-editor it doesn't seem likely for the article-editor to be unaware of prosemirror doc and its parts.

Yeah, article-editor is a different story which is not what I'm trying to change here, and I don't see a reason why we should change how it works. I do think we should move the adapt logic to body-editor, but it will still be used in article-editor.

----

So is the only problem you have here the usage of  `encode` on individual nodes? If so that can be changed.

mbartenev-atypon replied:
"That's fair. You're saying that encode might need data from somewhere else in the document, not that encode might update the nodes."
- that's right

mnatsheh commented:
Based on a discussion with @mbartenev-atypon I've moved the modals to the body-editor and changed it to use node attrs instead. It's still incomplete but please let me know if there is a problem with the approach.

mbartenev-atypon commented on src/views/bibliography_element.ts:
```diff
@@ -203,12 +199,12 @@ export class BibliographyElementBlockView<
   private handleSave = (item: BibliographyItem) => {
-    const node = this.decoder.decode(item) as BibliographyItemNode
-    this.updateNodeAttrs(node.attrs)
+    const node = decode(item)
```
I suppose in the final version there'd be no need to decode as we'll just receive the attributes, right?

mnatsheh replied:
Right. I reverted most of my changes to this file, _but the decode is still there_ since it was there before. Once this PR is merged, I'll do the same for references/citations and remove the decode.

mbartenev-atypon commented on src/views/citation_editable.ts:
```diff
@@ -137,11 +136,11 @@ export class CitationEditableView extends CitationView<EditableBlockProps> {
   private handleSave = (item: BibliographyItem) => {
-    if (item._id && !this.findPosition(item._id)) {
-      this.insertBibliographyNode(item)
+    const node = decode(item)
```
same here

mbartenev-atypon commented:
Besides usage of "encode" looks good.

mnatsheh commented:
This should be ready for review.
